Fix/add acknowledgment documentation (#85)
morpho-org/universal-rewards-distributor: README.md

# Universal Rewards Distributor
The Universal Rewards Distributor (URD), is a smart contract that allows for the distribution of multiple ERC20 tokens from a single offchain computed Merkle tree.

Each URD contract has an owner and a group of updaters (chosen by the owner). Values submitted by updaters are timelocked and can be revoked by 
the owner or overriden by another updater. However, this timelock can be set to 0 if the URD owner does not need it.

## Use Case Example

- Assume the Owner is a DAO with a periodic rewards mechanism. Each month, a [Gelato](https://www.gelato.network/) bot runs a script creating a Merkle tree that distributes TokenA and TokenB.

>   When setting up the URD as a DAO, I configure the timelock based on the risk of updater corruption (let's say 3 days), and add a Gelato bot as an updater. If I already have a distribution at the time of the URD deployment, I can define an initial root, or use an empty root if I don't.

Reviewer: do we keep the "I" then?

> # Universal Rewards Distributor
> The Universal Rewards Distributor (URD), is a smart contract that allows for the distribution of multiple ERC20 tokens from a single offchain computed Merkle tree.

Reviewer 2: ```suggestion
# Universal Rewards Distributor

The Universal Rewards Distributor (URD), is a smart contract that allows for the distribution of multiple ERC20 tokens from a single offchain computed Merkle tree.
```

Each URD contract has an owner and a group of updaters (chosen by the owner). Values submitted by updaters are timelocked and can be revoked by 

> the owner or overriden by another updater. However, this timelock can be set to 0 if the URD owner does not need it.

Reviewer 2: weird line jump ^^

# Universal Rewards Distributor
The Universal Rewards Distributor (URD), is a smart contract that allows for the distribution of multiple ERC20 tokens from a single offchain computed Merkle tree.

> Each URD contract has an owner and a group of updaters. Values of updaters are timelocked and can be revoked by 

Reviewer 3: ```suggestion
Each URD contract has an owner and a group of updaters (chosen by the owner). Values submitted by updaters are timelocked and can be revoked by 
```

> the owner or overrided by another updater. However, this timelock can be set to 0 if the URD owner do not need it.

Reviewer 3: ```suggestion
the owner or overriden by another updater. However, this timelock can be set to 0 if the URD owner does not need it.
```

> The Universal Rewards Distributor (URD), is a smart contract that allows for the distribution of multiple ERC20 tokens from a single offchain computed Merkle tree.

Reviewer: ```suggestion
The Universal Rewards Distributor (URD), is a smart contract allowing the distribution of multiple ERC20 tokens from a single offchain computed Merkle tree.
```

> The updaters can submit a new root to a pending value. Once the timelock period ends, anyone can accept this pending value. However, during the timelock period, the owner has veto power and can remove the pending value.

Reviewer 3: This feels like a dupplicate explanation, very similar to the explanation above

## Use Case Example

- Assume the Owner is a DAO with a periodic rewards mechanism. Each month, a [Gelato](https://www.gelato.network/) bot runs a script creating a Merkle tree that distributes TokenA and TokenB.

>   When setting up the URD as a DAO, I configure the timelock based on the risk of updater corruption (let's say 3 days), and add a Gelato bot as an updater. If I already have a distribution, I can define an initial root, or use an empty root if I don't.

Reviewer 3: ```suggestion
  When setting up the URD as a DAO, I configure the timelock based on the risk of updater corruption (let's say 3 days), and add a Gelato bot as an updater. If I already have a distribution at the time of the URD deployment, I can define an initial root, or use an empty root if I don't.
```

- It is possible to create a URD with no root, owner, or updaters. While this might seem pointless, it is the URD creator's responsibility to configure the URD correctly.

## Updaters Specifications

- Multiple updaters can be defined for a single URD. The owner can add or remove updaters at any time, instantly, without a timelock.
- All updaters share a single pending value. This means they can override the pending value (if any) at any time.
- If a pending value is not used as the main root, any submissions by updaters will override the pending value.

Having multiple updaters can lead to situations where the pending values are subject to multiple concurrent propositions. The owner must manage this scenario to maintain the URD's functionality. We recommend not having different reward distribution strategies in a single URD. All updaters should agree on a distribution mechanism and a root. The use case for having multiple updaters is to enhance resilience and security.

## Considerations for Merkle Tree

> - The claimable amount for a given user must always exceed the amount provided in the previous Merkle tree. If a claimer has claimed an amount higher than the amount in the Merkle tree (if claimed from a previous root), the claim will revert with a "root misconfigured" error.

Reviewer 3: ```suggestion
- The claimable amount for a given user must always exceed the amount provided in the previous Merkle tree. If a claimer has claimed an amount higher than the `claimable` amount in the Merkle tree (if claimed from a previous root), the claim will revert with a "root misconfigured" error.
```

# Universal Rewards Distributor

> The Universal Rewards Distributor (URD), is a smart contract that allows for the distribution of multiple ERC20 tokens from a single off-chain computed Merkle tree.

Reviewer: ```suggestion
The Universal Rewards Distributor (URD), is a smart contract that allows for the distribution of multiple ERC20 tokens from a single offchain computed Merkle tree.
```

> Each URD contract has an owner and a group of updaters. These updaters can submit a new root to a pending value, which is subject to a timelock (that can be set to 0). Once the timelock period ends, anyone can accept this pending value. However, during the timelock period, the owner has veto power and can remove the pending value.

Reviewer 2: I think the timelock should be introduced first, to avoid making nested sentences like "which is subject to a timelock (that can be set to 0)"
Nested/parentheses sentences are harder to read than simple, linear sentences

## Use Case Example

> - Assume I am a DAO with a periodic rewards mechanism. Each month, a [Gelato](https://www.gelato.network/) bot runs a script creating a Merkle tree that distributes TokenA and TokenB.

Reviewer: ```suggestion
- Assume that a DAO with a periodic rewards mechanism. Each month, a [Gelato](https://www.gelato.network/) bot runs a script creating a Merkle tree that distributes TokenA and TokenB.
```

Weird to use the first person, especially for a DAO

Author: User impersonating but I agree in some sense.

But your suggestion is not correct: "Assume that a DAO with a periodic rewards mechanism"

  When setting up the URD as a DAO, I configure the timelock based on the risk of updater corruption (let's say 3 days), and add a Gelato bot as an updater. If I already have a distribution, I can define an initial root, or use an empty root if I don't.

> - Each month, the Gelato bot proposes a new root. For 3 days, I have the opportunity to run checks on this root. After these 3 days, anyone can accept this value.

Reviewer: ```suggestion
- Each month, the Gelato bot proposes a new root. For 3 days, I have the opportunity to run checks on this root. After these 3 days and if I did not revoke the root, anyone can accept this value.
```

Author: not I, the DAO.

- The DAO must transfer the correct amount of tokens to the URD to allow all claimants to claim their rewards. If the DAO does not provide enough funds, the claim function will fail with the message "not enough funds in URD."

## Attaching an IPFS Hash

> - Using a Merkle tree delegates all root computation off-chain, leaving no information about the tree on-chain (except for the root). This makes it challenging for an integrator (or a claimer) to understand the Merkle tree or to know which proof to use. To address this, each root can be linked to an IPFS hash, which can link to any data. We recommend that all users follow the same format for the IPFS hash to facilitate integration. The suggested format is as follows:

Reviewer: ```suggestion
- Using a Merkle tree delegates all root computation offchain, leaving no information about the tree onchain (except for the root). This makes it challenging for an integrator (or a claimer) to understand the Merkle tree or to know which proof to use. To address this, each root can be linked to an IPFS hash, which can link to any data. We recommend that all users follow the same format for the IPFS hash to facilitate integration. The suggested format is as follows:
```

```json
{

>   "id": "A string id of the merkle tree, can be random",

Reviewer 2: ```suggestion
  "id": "A string id of the Merkle tree, can be random",
```

  "metadata": {

>   "info": "a key value mapping allowing you to add information"

Reviewer 2: ```suggestion
    "info": "a key value mapping allowing you to add information"
```

  },
  "root": "The merkle root of the tree",
  "tree": [
       {
              "account": "The address of the claimer",
              "reward": "The address of the reward token",
              "claimable": "The claimable amount as a big number",
              "proof": ["0x1...", "0x2...", "...", "0xN..."]

>       }

Reviewer 2: weird formatting? indentation was 2 then here it's 4

  ]
}
```

- We recommend not including spaces or new lines when uploading to IPFS, to ensure a consistent method of uploading your JSON file.

## Owner Specifications

> - The URD is an owner-managed contract, meaning the owner has full control over its distribution. Specifically, the owner can bypass all time-locked functions, modify the time-lock, add or remove updaters, and revoke the pending value at their discretion.

Reviewer: ```suggestion
- The URD is an owner-managed contract, meaning the owner has full control over its distribution. Specifically, the owner can bypass all timelocked functions, modify the timelock, add or remove updaters, and revoke the pending value at their discretion.
```

> - If the owner is set to a zero address, the contract becomes ownerless. In this scenario, only updaters can submit a root. Furthermore, the time-lock value becomes unchangeable, and pending values cannot be removed.

Reviewer: ```suggestion
- If the owner is set to the zero address, the contract becomes ownerless. In this scenario, only updaters can submit a root. Furthermore, the time-lock value becomes unchangeable, and pending values cannot be removed.
```

> - If there are neither owner nor updaters, the URD becomes immutable. This is useful for creating a one-time distribution that must remain unchanged. This can be accomplished at the contract's creation by providing only a Merkle root and an optional IPFS hash. After this, the contract's sole functionality is to claim rewards.

Reviewer: The URD is already immutable, when funds are deposited and the root has been submitted, it's trustless to me

Author: Not really imo, owner can still rug users. So how define a contract as trustless if someone can rug you?

Reviewer: how?

Author: Oh ok got it, changing it

- It is possible to create a URD with no root, owner, or updaters. While this might seem pointless, it is the URD creator's responsibility to configure the URD correctly.

## Updaters Specifications

- Multiple updaters can be defined for a single URD. The owner can add or remove updaters at any time, instantly, without a timelock.
- All updaters share a single pending value. This means they can override the pending value (if any) at any time.
- If a pending value is not used as the main root, any submissions by updaters will override the pending value.

> Having multiple updaters can lead to situations where the pending values are subject to multiple concurrent propositions. The owner must manage this scenario to maintain the URD's functionality. We recommend not having different reward distribution strategies in a single URD. All updaters should agree on a distribution mechanism and a root.The use case for having multiple updaters is to enhance resilience and security.

Reviewer: ```suggestion
Having multiple updaters can lead to situations where the pending values are subject to multiple concurrent propositions. The owner must manage this scenario to maintain the URD's functionality. We recommend not having different reward distribution strategies in a single URD. All updaters should agree on a distribution mechanism and a root. The use case for having multiple updaters is to enhance resilience and security.
```

## Considerations for Merkle Tree

- The claimable amount for a given user must always exceed the amount provided in the previous Merkle tree. If a claimer has claimed an amount higher than the amount in the Merkle tree (if claimed from a previous root), the claim will revert with a "root misconfigured" error.

> - TODO: Define the list of invariants for a new root.

Reviewer: todo

Author: Yes I know, imo it is not in the scope of this PR. I have to think more in-depth about all the checks we can do on a specific Merkle tree.
